closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -47,6 +48,47 @@
 # include <pthread.h>
 #endif /* defined(WINDOWS) */
 
+/*
+ * This definition is needed to compile the "OSSL_RSA_get0_key_compatibility" function
+ * Since the first parameter of that function is of type RSA, which is
+ * a typedef of rsa_st, the function would not compile without defining
+ * rsa_st here */
+struct rsa_st {
```

Comment:
I still object to this declaration. We should not assume anything about the definition of what is intended to be an opaque type.

Reply:
Further, this appears to be a copy of https://github.com/openssl/openssl/blob/OpenSSL_1_1_1k/crypto/rsa/rsa_local.h#L28-L66, but I see no mention of the openssl licenses in this file.

Reply:
@keithc-ca The issue is, I don't think there's a way around it. I've been trying to look at other functions in OpenSSL 1.0.* that would return the same fields as `RSA_get0_key` in OpenSSL 1.1.*, but I couldn't find any. And this makes sense because `rsa_st` was not opaque and those fields could be accessed directly without the need for functions.

In [OpenSSL's wiki](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes), it is mentioned that one of the API changes in 1.1.0 as compared to previous versions is that the structs are now opaque. The backward compatibility issue was also expected and the the suggested solution for that was to make those compatibility functions. Now if we compile the JDK with OpenSSL 1.0.*, we won't need to add this structure definition since it is available in an included header file ([rsa.h](https://github.com/openssl/openssl/blob/OpenSSL_1_0_0-stable/crypto/rsa/rsa.h#L132-L165)). However, in OpenSSL 1.1.*, the structure is not defined in any included header file and is not expected to be used. Hence, the only way that I could think of to support OpenSSL 1.0.* even when compiling the JDK with OpenSSL 1.1.* is to add the definition itself there.

In terms of the `lock` field, you're right. I took that from OpenSSL 1.1.1 by mistake. I removed it in my latest commit

Reply:
After discussing this more with Keith and Alon on slack, I solve this issue as follows:
I remove the `rsa_st` definition from our native source code and fall back on the pure java implementation when necessary.

In pseudocode:

```java
if (openssl version >= 1.1.*) {
  keep using our new code to verify
} else {
  if (key has a size that would cause an error in the old native implementation) {
    use pure java implementation
  } else {
    use OSSL_RSA_public_decrypt to verify
  }
}
```